### ISSUE TITLE
[Windows 10] [Nvidia GTX 1650] Flycast Crashing When Closing Content/Launching Games Outside of Core

### ISSUE THREAD

**Issue by ds22x** [libretro/flycast#974]
As previously discussed [here](https://forums.libretro.com/t/flycast-crashing-when-closing-content-launching-games-outside-of-core/29862), Flycast currently crashes RetroArch when closing content or launching a game associated with another core, unfortunately without providing a crash log of any kind, regardless of whenever the threaded renderer is used or not, video driver used or any other settings such as resolution and enhancements and hacks.

After some further investigation, it appears to be related to a Game Ready driver update for the Nvidia GTX 1650 GPU (specifically 452.06 and above), as anything before this will have Flycast not crash upon shutting down the core, as will using the build-in Intel UHD Graphics 630 of my Intel Core i5-9300H.
Also went and tried out several older builds of the Flycast core, going as far back as [df4d2d98f03de8b235c2bb01fccc078b36fcb12b](https://github.com/libretro/flycast/tree/df4d2d98f03de8b235c2bb01fccc078b36fcb12b), and the aforementioned problem still persists.

The full specs of my computer are as followed:
- Windows 10 build 2004
- Intel Core i5-9300H @2,40GHz
- Nvidia GeForce GTX 1650
- 16GB DDR4 RAM @2667MHz
- RetroArch version 1.90

**Comment by barbudreadmon** [comment]
> specifically 452.06 and above

Meaning https://github.com/libretro/flycast/issues/968 isn't the only regression from those recent nvida drivers ? Or are you maybe using vulkan with 4GB pixel buffer ? If so, using 2GB fixes this ?

**Comment by ds22x** [comment]
Used both Vulkan and GL, set it to any of the pixel buffer settings, as well as set it to "per strip" and "per triangle" options, and the problem still occurs

**Comment by afjara** [comment]
Also crashing for me. Tried to use an older build from july 2020 and the problem persists. I recently updated Windows 10 to 20H2. Using latest nvidia drivers from 2020.10.7 with my Core i3-7130U, GeForce MX130 laptop.

I will check if the problem is present when trying with an Intel HD Graphics 610 laptop and a GTX950 desktop. Flycast settings set to default, RetroArch video driver set to glcore.

**Comment by afjara** [comment]
> 
> 
> Also crashing for me. Tried to use an older build from july 2020 and the problem persists. I recently updated Windows 10 to 20H2. Using latest nvidia drivers from 2020.10.7 with my Core i3-7130U, GeForce MX130 laptop.
> 
> I will check if the problem is present when trying with an Intel HD Graphics 610 laptop and a GTX950 desktop. Flycast settings set default, RetroArch video driver set to glcore.

Flycast **DOES NOT crash** with my **Intel HD Graphics 610 laptop** neither my **GeForce GTX950 desktop** (also updated to latest 456.71 nvidia driver). All my systems are running Windows 10 20H2, RetroArch with glcore video driver, flycast at default settings.

Only crashing system is my MX130 laptop.

**Comment by barbudreadmon** [comment]
That's gonna be a real tough nut to crack if it only happens on certain models, my gtx960 doesn't seem to produce this crash either. We'll need some backtrace from someone with a faulty model.

nvidia seriously screwed up there...

**Comment by marcrobledo** [comment]
Happening to me with gtx970 aswell.

**Comment by marcrobledo** [comment]
~~Just updated the core and it's not crashing anymore after changing game!~~
Nevermind. It happened again.

**Comment by ds22x** [comment]
Still happens on my setup with the latest drivers.

**Comment by ds22x** [comment]
Still happens with the latest Nvidia drivers, and also happens with the most recent PCSX2 core.

**Comment by Immersion95** [comment]
I also have the 3070 card with latest drivers and it's crashing. I will try to get a backtrace tomorrow.

**Comment by Alextended** [comment]
Same here, GTX1080/i73770K.

**Comment by marcrobledo** [comment]
I updated to 20caf3f, and looks like it's able to swap games without crashing at last.

**Comment by ds22x** [comment]
Swap Dreamcast games works fine sure, but unloading Flycast is what still crashes.

**Comment by marcrobledo** [comment]
> 
> 
> Swap Dreamcast games works fine sure, but unloading Flycast is what still crashes.

Works fine here (GTX 970).

**Comment by voorunde** [comment]
It took a while to figure out but I finally discovered that Flycast crashes upon exiting content if Retroarch's global video driver is set to Vulkan, it doesn't happen with gl/glcore. This happens regardless of whether Flycast's driver is set Vulkan or GL/GLcore. Flycast also crashed for me for having devreorder's dinput8.dll in the Retroarch root folder. Once I deleted dinput8.all and set Retroarch's driver to glcore it stopped crashing and terminated as expected.

**Comment by ds22x** [comment]
~~While the above didn't fix it for me, updating to the latest nightly of RetroArch and Nvidia's 461.92 drivers did!
Works with both GL and Vulkan as well.~~

**Comment by bslenul** [comment]
Oh interesting, I tried voorunde suggestion, changing to glcore globally, then creating a vulkan override for the core and yeah, no crash anymore on close content. Updating to latest Nvidia drivers didn't fix it for me however.

Steps to reproduce if anyone is interested:

* No content loaded, set video driver to glcore.
* Restart RetroArch (or else the "fix" won't work properly).
* Start a DC game, change driver to vulkan and create a core override.
* Close content.
* Launch the game again, it should be using vulkan and it shouldn't crash anymore on close content.

Another fix for me was to downgrade my Nvidia drivers to 451.48 (version mentioned by Tatsuya on Discord) using the "clean install" option or whatever it's called.

Tested by running RetroArch from MinGW, because for some weird mystical reasons if I launch RA from Windows Explorer and if I get Flycast to crash a couple times it doesn't crash anymore but I'm getting a ton of weird stuff: https://github.com/libretro/RetroArch/issues/11940

**Comment by ds22x** [comment]
Well apparently it only 'fixed' itself because Windows decided to add a register key that would prevent the core from unloading when RetroArch runs...
Without it, it will crash on close content, regardless of video driver.

**Comment by Immersion95** [comment]
I updated to the latest Nvidia driver from March, 30th, latest bios (which adds Resize Bar), latest Windows update and latest RA+Core nightlies from today and the issue seems to be resolved.

I'm in Vulkan globally fyi.

**Comment by ds22x** [comment]
Have you tried loading up another Dreamcast game after closing the first one and see if the second game behaves incorrectly?
Because Windows 10 likes to sometimes add a registry key when Flycast crashes too many times, which will prevent the Flycast core .dll from unloading, which 'fixes' the crashing, but introduces a whole new list of bugs, including a second Dreamcast game not loading properly, to stuff like random crashes when trying to load another core.

**Comment by Immersion95** [comment]
> Have you tried loading up another Dreamcast game after closing the first one and see if the second game behaves incorrectly?
> Because Windows 10 likes to sometimes add a registry key when Flycast crashes too many times, which will prevent the Flycast core .dll from unloading, which 'fixes' the crashing, but introduces a whole new list of bugs, including a second Dreamcast game not loading properly, to stuff like random crashes when trying to load another core.

Tested everything, no problem whatsoever !

**Comment by ds22x** [comment]
Bit of a sanity check, but if you open Regedit, and navigate to [HKEY_CURRENT_USER\SOFTWARE\Microsoft\Windows NT\CurrentVersion\AppCompatFlags\Layers], do you see an entry with the data description "$IgnoreFreeLibrary<flycast_libretro.dll>"?

**Comment by Immersion95** [comment]
> Bit of a sanity check, but if you open Regedit, and navigate to [HKEY_CURRENT_USER\SOFTWARE\Microsoft\Windows NT\CurrentVersion\AppCompatFlags\Layers], do you see an entry with the data description "$IgnoreFreeLibrary<flycast_libretro.dll>"?

Nothing :).

Is the bug fixed for other people ?

**Comment by akumenon** [comment]
I started using RA on October '20 or so, and I've always had this problem, regardless of video driver, threaded rendering setting, GPU drivers for GTX 970. Haven't tried other GPUs yet. I don't have the registry entry mentioned above. I'll be trying this out on other GPUs when possible.

**Comment by Alextended** [comment]
If I change RA to glcore and not customize the cores back to Vulkan, it works without issue for me. If I change Flycast itself back to Vulkan however, it appears to work as it doesn't crash when I close a game, but it will crash when I try to load another core's game. So it's basically the same thing, you could load different Flycast games before if you didn't try to close the current first.

Oh well, it's all good with glcore! At the moment at least. GTX1080/latest Win10/updated everything, drivers included.

**Comment by bslenul** [comment]
No idea if one of the backports fixed it or if it's really by updating the Nvidia drivers, but I don't have the issue anymore, no crash on close content, no registry key created anymore, no weird behavior on next loads, etc.

**edit:** downgraded both the core and RetroArch, still no issue so I guess the Nvidia drivers fixed it indeed! Thanks for the heads up @Immersion95 Also not sure if it matters but I checked the "make a clean install" box while updating drivers.

**Comment by ds22x** [comment]
Using latest drivers (465.89) with up to date Flycast (0.1 de6ccff) and RetroArch (1.9.1), and it still crashes on close content for me.

**Comment by Immersion95** [comment]
> Using latest drivers (465.89) with up to date Flycast (0.1 [de6ccff](https://github.com/libretro/flycast/commit/de6ccff686c06e63ccf7ebf08332e72031fd30db)) and RetroArch (1.9.1), and it still crashes on close content for me.

Latest RA nightlies ? Maybe it's only fixed for Ampere....

**Comment by bslenul** [comment]
> Maybe it's only fixed for Ampere....

No, I'm still using a GTX 970 and it started to work properly again today after updating Nvidia drivers. @ds22x have you checked the "clean installation" during drivers update? Like I said I checked it, but it could be completely unrelated... worth the shot I guess?

Could be something else tho, like a Windows update? 🤷

**Comment by ds22x** [comment]
Checked latest nightly, still crashes.
Did clean install those latest drivers after first reverting back to default, still no dice.
Still using a GTX 1650 and running W10 20H2.

**Comment by bslenul** [comment]
So weird, in my case it's clearly the drivers that fixed it, I just downgraded to 461.92 and the crashes are back, re-installed the newest 465.89 and no crash or anything weird anymore.

I also noticed there was a Geforce Experience update, are you using it? Is it up-to-date (3.22.0.32)?

And while typing this I just got a notification saying a new version for the drivers (466.11) is available :p

**Comment by ds22x** [comment]
Installing the latest Geforce Experience update doesn't fix it for me either unfortunately.
I'll try and see if clean installing 466.11 (when it's available for me) does anything.

EDIT: Just clean installed 466.11, and still a crash after closing content on both stable and nightly.

**Comment by ds22x** [comment]
Just discovered something very interesting.
If I only use an external display with my laptop, it doesn't crash!
But if have the laptop screen in use in any way however, it crashes on close content.

**Comment by spikeyxx** [comment]
I am having the same problem, here is my setup. If I can supply any more info that will be of use please let me know.

GTX 1070 Ti,
Windows 10, version 20H2, Up to date
New Install of RetroArch 1.9.1 with latest Flycast Core Update.
Clean installation of Driver 466.11

UPDATE -

Made another fresh installation of RetroArch, seems to be solved now. Not sure what fixed it.

**Comment by ds22x** [comment]
Are you using a laptop by any change?

**Comment by spikeyxx** [comment]
No. TV

> 
> 
> Are you using a laptop by any change?

No, TV

**Comment by ds22x** [comment]
> Just discovered something very interesting.
> If I only use an external display with my laptop, it doesn't crash!
> But if have the laptop screen in use in any way however, it crashes on close content.

Just to add to this, but if I have my main menu video driver set to D3D11, and with a core video driver override for either GL or Vulkan, it will also crash on close content even with the external display setup.
If I however set the main menu video driver to GL and the core video driver override to Vulkan or visa verse, it doesn't crash (that is, with the external display setup)

**Comment by spikeyxx** [comment]
There's something odd going on with retroarch crashing when changing
content. I've made another topic and I'm hoping it gets some attention.
It's been going on for years and I think it has more to do with the program
itself than cores.

**Comment by PoloniumRain** [comment]
I have this problem with both a GTX 980 Ti and GTX 1080 Ti, and the latter is also on a clean install of Windows 10 with all the latest updates.

I've also tried:
Older versions of Windows 10 (all versions since early 2020)
Clean installs of Retro Arch 1.9.0 to 1.9.4
Latest RA nightly build
Multiple different Nvidia drivers including the latest ones (currently 466.47)
GL, GLcore, DX11 and Vulkan display drivers

Crashing has always been a problem with Flycast since i started using it at least a year ago. It's not Nvidia drivers causing the crashes and it's not any Windows updates, and if it is, it's not anything from the past 12 months (or more).

Lately Flycast wont crash when closing a game, but then when i start any game that uses a certain core, like Dolphin, PPSSPP, Beetle PSX HW, or DuckStation, it will then crash. That's always been a problem though, but most of the time it would just crash when closing a game so i couldn't even get to try another core/game. I'm pretty sure this has happened before as well, where it stopped crashing when closing games, but then it came back. So i have my doubts when people think they've fixed this issue. 
One thing also worth noting is that the DX11 display driver always still crashes on closing, but not Vulkan, GL or GLcore.

So to anyone who thinks they've fixed it, can you run a game with Flycast, close it, then run any games that use Dolphin, PPSSPP, Beetle PSX HW, or DuckStation? Does it crash for you?


> Have you tried loading up another Dreamcast game after closing the first one and see if the second game behaves incorrectly?

@ds22x I also had this problem and it looks like a separate bug... Because (bizarrely) i seem to have fixed it by removing all the spaces from the names of directories/folders that the Retro Arch install is located inside. For example, my install folder was named "Retro Arch" and i changed it to "RetroArch" (no space). I've tested this many times now and it's solved this specific problem, but the crashing still happens.

**Comment by ds22x** [comment]
My installation folder is literally C:/emulators/retroarch/, so that ain't it.

**Comment by PoloniumRain** [comment]
Weird... It works for me so i've no idea.

Mine is on E:\ drive, but i just tried a clean install on C:\ drive using the same location as you to see if it's the drive letter, and it works using that location without spaces, but when i change it to C:\Emulators\Retro Arch\ it wont work again. Specifically, after closing a game and starting another game, Flycast will start the _previous_ game or get stuck in a boot loop or black screen. No other cores do this.

**Comment by ds22x** [comment]
That sounds more like the Flycast core and its loaded content don't get unloaded properly, which is similar to the problem caused by the registry entry that gets added in Windows 10 when Flycast crashes, which prevents the Flycast .dll from unloading, causing similar issues.

**Comment by bslenul** [comment]
Yeah if it works after renaming then it means you probably have the registry key: open regedit and go to `HKEY_CURRENT_USER\SOFTWARE\Microsoft\Windows NT\CurrentVersion\AppCompatFlags\Layers`, if you see a `$ IgnoreFreeLibrary<flycast_libretro.dll>` key, just delete it and it should have the same behavior as before.

The Nvidia drivers fixed the main issue for me: it doesn't crash on close content anymore, and I don't get any weird behavior when launching multiple DC games in a row.

However you're completely right about switching to a different core (can't believe I've never tried that before...), I tried Flycast -> close content -> PPSSPP/DuckStation/PCSX2 and got a crash every time (but no issue with Beetle PSX HW for some reason), and it re-created the registry key after a couple crashes :/

**Comment by spikeyxx** [comment]
I had similar behaviour on Mupen, where it wouldn't load a new game after
closing another game unless I changed the name of retroarch.exe, or changed
the folder it was in.

I think it's a Retroarch bug, not a core bug.

Hunterk thought he had heard somebody else talking about a similar issue,
but couldn't find anything when he asked around.

**Comment by PoloniumRain** [comment]
> Yeah if it works after renaming then it means you probably have the registry key: open regedit and go to `HKEY_CURRENT_USER\SOFTWARE\Microsoft\Windows NT\CurrentVersion\AppCompatFlags\Layers`, if you see a `$ IgnoreFreeLibrary<flycast_libretro.dll>` key, just delete it and it should have the same behavior as before.

Ah yeah i had the key. Deleted it and that seems to have fixed it for now...  @spikeyxx Maybe this is also causing the problem you mentioned in your last post?

> The Nvidia drivers fixed the main issue for me: it doesn't crash on close content anymore, and I don't get any weird behavior when launching multiple DC games in a row.

@bslenul Did you try the Retro Arch D3D11 video driver? That still crashes on close with a 980 Ti and 1080 Ti (latest drivers). [Similar situation](https://github.com/libretro/flycast/issues/974#issuecomment-840922554) for @ds22x as well.

> However you're completely right about switching to a different core (can't believe I've never tried that before...), I tried Flycast -> close content -> PPSSPP/DuckStation/PCSX2 and got a crash every time (but no issue with Beetle PSX HW for some reason), and it re-created the registry key after a couple crashes :/

Yeah it re-created the registry key for me as well. And found another that crashes - SwanStation. 

This thread is pretty long, so just to sum things up so far:

- The Flycast crashing issue has been around for a very long time (years?).
- Recent Nvidia drivers fix the crashing on close for some people.
- But the D3D11 video driver always still crashes on close for everyone (so far).
- Still crashes for everyone (so far) when closing Flycast and then running specific cores: 
- (DuckStation/SwanStation/Beetle PSX HW/PCSX2/PPSSPP/Dolphin)
- Crashing might be Nvidia specific. No mention of AMD/Intel crashing so far.
- Looks like everyone with crashing is using Windows 10.
- The Windows registry key is due to RA crashing and is at least the reason for _some_ other weird behaviour.

Does this sound about right?

**Comment by bslenul** [comment]
> Did you try the Retro Arch D3D11 video driver?

Yeah crashes too, which is weird because Flycast doesn't have a D3D11 renderer, so it automatically switches to Vulkan on launch... 🤔

**Comment by spikeyxx** [comment]
What are these registry flags? I have two. One for flycast and another for
pcsx2

**Comment by ds22x** [comment]
Forgot to post it here, but https://github.com/flyinghead/flycast/commit/de060d66ab1afd319b6290d58f0c58659db55043 fixed this issue, although for the upstream Libretro core (hence why I haven't closed this issue report yet).
